Multisig/UI/App/UIKit/Transaction/ActionDetailViewController.swift

```diff
@@ -0,0 +1,309 @@
+        cell.setText(text, style: GNOTextStyle.body.color(.gnoDarkGrey))
+        cell.onTap = {
+            Self.copyValue(text)
+        }
+        cell.margins.leading += indentation
+        return cell
+    }
+
+    private func emptyCell(indentation: CGFloat = 0) -> UITableViewCell {
+        let cell = tableView.dequeueCell(ActionDetailTextCell.self)
+        cell.setText("empty", style: GNOTextStyle.body.color(.gnoMediumGrey))
+        cell.selectionStyle = .none
+        cell.margins.leading += indentation
+        return cell
+    }
+
+    private func addressCell(_ address: Address, indentation: CGFloat = 0) -> UITableViewCell {
+        let cell = tableView.dequeueCell(ActionDetailAddressCell.self)
+        cell.setAddress(address)
+        cell.selectionStyle = .none
+        cell.margins.leading += indentation
+        return cell
+    }
+
+    private func hexCell(_ text: String, indentation: CGFloat = 0) -> UITableViewCell {
+        let cell = self.tableView.dequeueCell(ActionDetailTextCell.self)
+        cell.setText(text, style: .body)
+        cell.onTap = {
+            Self.copyValue(text)
+        }
+        cell.margins.leading += indentation
+        return cell
+    }
+
+    /// Creates expandable cell that produces other cells when expanded.
+    ///
+    /// - Parameters:
+    ///   - text: Text of the expandable header
+    ///   - indentation: indentation width
+    ///   - content: closure that produces new cells to insert right after
+    ///     the expanded cell.
+    private func expandableCell(_ text: String,
+                                indentation: CGFloat = 0,
+                                content: [UITableViewCell]) -> UITableViewCell {
+        let cell = tableView.dequeueCell(ActionDetailExpandableCell.self)
+        cell.setText(text)
+        cell.state = .collapsed
+        cell.margins.leading += indentation
+        cell.subcells = content
+
+        cell.onTap = { [weak self, weak cell] in
+            guard let `self` = self, let cell = cell, let cellIndex = self.cells.firstIndex(of: cell), !cell.subcells.isEmpty else { return }
+
+            switch cell.state {
+            case .collapsed:
+                // then expand the contents
+
+                cell.state = .expanded
+
+                // create cells
+                let insertionIndex = cellIndex.advanced(by: 1)
+                self.cells.insert(contentsOf: cell.subcells, at: insertionIndex)
+
+                // update UI
+                self.tableView.beginUpdates()
+                let insertedPaths = (insertionIndex..<insertionIndex.advanced(by: cell.subcells.count)).map { IndexPath(row: $0, section: 0) }
+                self.tableView.insertRows(at: insertedPaths, with: .bottom)
+                self.tableView.endUpdates()
+
+            case .expanded:
+                // then collapse the contents
+                cell.state = .collapsed
+
+                // visit all subcell tree and collapse all
```

Comment:
Not sure, but from UX perspective I think we can keep the subcells expanded and just collapse the parent.

Reply:
agree, but it'll be more complicated to implement, I'm leaving this as is